src/main/java/org/skife/jdbi/v2/sqlobject/ParameterBinderRegistry.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.skife.jdbi.v2.sqlobject;
+
+import java.lang.reflect.Method;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+class ParameterBinderRegistry
+{
+    static final ParameterBinderFactory DEFAULT_BINDER_FACTORY = new ParameterBinderFactory() {
+        @Override
+        public boolean accepts(Class type, Method method, int param_idx) {
+            return true;
+        }
+
+        @Override
+        public Binder binderFor(Class type, Method method, int param_idx) {
+            return new PositionalBinder(param_idx);
+        }
+    };
+
+    private final List<ParameterBinderFactory> orderedBinderFactories = new CopyOnWriteArrayList<ParameterBinderFactory>();
+    private final ConcurrentHashMap<ClassMethodParameter, ParameterBinderFactory> factoryCache = new ConcurrentHashMap<ClassMethodParameter, ParameterBinderFactory>();
+
+    private static class ClassMethodParameter {
+        final Class type;
+        final Method method;
+        final int param_idx;
+
+        ClassMethodParameter(Class type, Method method, int param_idx) {
+            this.type = type;
+            this.method = method;
+            this.param_idx = param_idx;
+        }
```

Comment:
I'm wondering if `type` is redundant considering that we already have `method.getDeclaringClass()`. (I'm mostly musing about v3 here)